bciers/apps/dashboard/middlewares/withTokenRefresh.ts

```diff
@@ -0,0 +1,118 @@
+   */
+  return Date.now() / 1000 > (token?.expires_at ?? 0);
+}
+
+async function fetchNewAccessToken(refreshToken: string | undefined) {
+  /**
+   * Attempt to refresh access token from the OAuth provider
+   */
+  const newTokensResponse = await fetch(`${process.env.KEYCLOAK_TOKEN_URL}`, {
+    method: "POST",
+    body: new URLSearchParams({
+      client_id: `${process.env.KEYCLOAK_CLIENT_ID}`,
+      client_secret: `${process.env.KEYCLOAK_CLIENT_SECRET}`,
+      grant_type: "refresh_token",
+      refresh_token: `${refreshToken}`,
+    }),
+  });
+
+  if (!newTokensResponse.ok) {
+    console.error(
+      `Failed to refresh access token from KC: ${newTokensResponse.status}`,
+    );
+    return undefined;
+  }
+
+  const newTokens = await newTokensResponse.json();
+
+  return {
+    access_token: newTokens.access_token,
+    refresh_token: newTokens.refresh_token,
+    expires_at:
+      Math.round(Date.now() / 1000) + OAUTH_TOKEN_ROTATION_INTERVAL_SECONDS,
+  };
+}
+
+function signOut(request: NextRequest) {
+  /**
+   * Signs out the user by redirecting to the sign-in page
+   * and clearing the session cookies.
+   */
+  const response = NextResponse.redirect(new URL("/onboarding", request.url));
+
+  request.cookies.getAll().forEach((cookie) => {
+    if (cookie.name.includes("next-auth")) response.cookies.delete(cookie.name);
+  });
+
+  return response;
+}
+
+export const withTokenRefreshMiddleware: MiddlewareFactory = () => {
+  /**
+   * Middleware that supplements next-auth to add access token rotation.
+   * Huge thanks to the community: https://github.com/nextauthjs/next-auth/discussions/9715#discussioncomment-12818495
+   */
+  return async (request: NextRequest) => {
+    // Casting to our augmented JWT type
+    const jwt = (await getToken({
+      req: request,
+      secret: process.env.NEXTAUTH_SECRET,
+    })) as JWT | null;
+
+    const response = NextResponse.next();
+
+    if (!jwt) return response;
+
+    if (shouldUpdateToken(jwt)) {
+      const newKcTokens = await fetchNewAccessToken(jwt.refresh_token);
+
+      if (!newKcTokens) {
+        // There was an error refreshing the tokens
+        return signOut(request);
+      }
+
+      const newSessionToken = await encode({
+        secret: `${process.env.NEXTAUTH_SECRET}`,
+        token: {
+          ...jwt,
+          ...newKcTokens,
+        },
+        maxAge: 30 * 60, // 30 min to match KC
+        salt: SESSION_COOKIE,
+      });
+
+      const size = 3933; // maximum size of each chunk
+      const regex = new RegExp(`.{1,${size}}`, "g");
+
+      // split the string into an array of strings
+      const tokenChunks = newSessionToken.match(regex);
+
+      if (tokenChunks) {
+        tokenChunks.forEach((tokenChunk, index) => {
+          response.cookies.set(`${SESSION_COOKIE}.${index}`, tokenChunk, {
+            httpOnly: true,
+            maxAge: 30 * 60, // 30 min to match KC
+            secure: SESSION_SECURE,
+            sameSite: "strict",
```

Comment:
@andrea-williams  am I going to run into trouble here?

Reply:
hmmm possibly? the test will be whether you're able to refresh your token again after your `tokenChunks` expire. And if you haven't already, I would recommend checking that the cookies are actually being set with sameSite=strict - next-auth likes to override what you specify most of the time in favour of their own (Lax) default.